### PR TITLE
Fedora 6 Demo 1

### PR DESCRIPTION
**Pull request by dbernstein**
**Make all fixes necessary to be able to perform the first Fedora 6 demo **
* * *

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3159

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
1. GET the repository root
2. PUT a basic container
3. GET a basic container

# What's new?
RepositoryInitializer  automatically  populates the FedoraToOCFLIndex with the root  repository node.
# How should this be tested?

```
mvn clean install -Pone-click
```

```
java  -Dfcrepo.ocfl.staging.dir=target/fcrepo-data/staging \ 
-Dfcrepo.ocfl.storage.root.dir=target/fcrepo-data/ocfl-root \ 
-Dfcrepo.ocfl.work.dir=target/fcrepo-data/work  \ 
-jar fcrepo-webapp/target/fcrepo-webapp-6.0.0-SNAPSHOT-jetty-console.jar --headless
```
```
# GET repo root
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest -i
# PUT a basic container
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/test -v  -H "Content-Type: text/turtle"  -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'My Test RDF'"
# GET a basic container
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/test -i
```

# Additional Notes:
It is no longer necessary to delete the ./fcrepo-data  and ./fcrepo-http-api/fcrepo-data directories between runs.  However you should specify your ocfl directories in the commandline as I have indicated above when running the one-click.  Additionally you should delete the ./target/fcrepo-data directory in between runs of the one-click.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo4/committers

### CODE REVIEW

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java [hunk 1]
```diff
@@ -55,7 +54,7 @@ private static final File resolveDir(final String systemPropertyKey, final Strin
             return new File(path);
         } else {
             //return default
-            return Paths.get(getProperty(JAVA_IO_TMPDIR), defaultDirectoryName).toFile();
+            return Paths.get("fcrepo-data", defaultDirectoryName).toFile();
```
Comment (bbpennel) [inline fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: Guessing we'll eventually want a more permanent location for this, probably from config? Maybe its not worthwhile to make a constant yet, hmm.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java [hunk 2]
```diff
@@ -173,7 +173,9 @@ public static WriteOutcome writeRDF(final OCFLObjectSession session, final RdfSt
         try (final var os = new ByteArrayOutputStream()) {
             final StreamRDF streamRDF = getWriterStream(os, getRdfFormat());
             streamRDF.start();
-            triples.forEach(streamRDF::triple);
+            if(triples != null) {
```
Comment (bbpennel) [inline fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java]: space.

Is this covering the case where a client provides an empty body for a rdf source?

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/RdfLexicon.java [hunk 3]
```diff
@@ -131,6 +134,9 @@
     public static final Resource FEDORA_PAIR_TREE =
             createResource(REPOSITORY_NAMESPACE + "Pairtree");
 
+    //RDF
+    public static final Node RDF_TYPE = createURI(RDF_NAMESPACE + "type");
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/RdfLexicon.java]: will fix.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/RepositoryInitializer.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    @Inject
+    private RdfSourceOperationFactory operationFactory;
+
+    /**
+     * Initializes the repository
+     */
+    @PostConstruct
+    public void initialize() {
+        //check that the root is initialized
+        final PersistentStorageSession session = this.sessionManager.getSession("initializationSession" + System.currentTimeMillis());
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/RepositoryInitializer.java]: will fix.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/api/FedoraToOCFLObjectIndex.java [hunk 5]
```diff
@@ -31,10 +31,10 @@
      * as an OCFL object and returns its identifiers.
      *
      * If you pass fedora identifier that is not part of an archival group such as
-     * "my/fedora/binary/fcr:metadata"  the parent fedora resource returned in the mapping will be "my/fedora/binary".
-     *
+     * "my/fedora/binary/fcr:metadata"  the  fedora resource returned in the mapping will be "my/fedora/binary".
+     * <p>
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/api/FedoraToOCFLObjectIndex.java]: Hmm. I still have them turned off.  Not sure why this one appeared.  Will remove.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/RepositoryInitializer.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        } catch (PersistentItemNotFoundException e) {
+            final RdfSourceOperation operation = this.operationFactory.createBuilder(FEDORA_ID_PREFIX, "http://www.w3.org/ns/ldp#BasicContainer")
+                    .parentId(FEDORA_ID_PREFIX).build();
+            try {
+                session.persist(operation);
+                session.commit();
+            } catch (PersistentStorageException ex) {
+                throw new RepositoryRuntimeException(e);
+            }
+        } catch (PersistentStorageException e) {
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/RepositoryInitializer.java]: will do.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java (hunk 1 above)
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: I'll put one in.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java (hunk 2 above)
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java]: I believe so, but I'll check.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java [hunk 7]
```diff
@@ -447,7 +439,7 @@ private RdfStream getTriples(final FedoraResource resource, final TripleCategory
 
     protected URI getUri(final FedoraResource resource) {
         try {
-            final String uri = translator().reverse().convert(resource).getURI();
+            final String uri = this.identifierConverter.toExternalId(resource.getId());
```
Comment (awoods) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: It seems safer to use `identifierConverter()` getter... instead of the `this.identifierConverter` directly.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java [hunk 8]
```diff
@@ -447,7 +439,7 @@ private RdfStream getTriples(final FedoraResource resource, final TripleCategory
 
     protected URI getUri(final FedoraResource resource) {
         try {
-            final String uri = translator().reverse().convert(resource).getURI();
```
Comment (awoods) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: As a separate issue, should we go ahead and purge all usage of `translator()`?

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java [hunk 9]
```diff
@@ -425,8 +428,8 @@ public Response createOrReplaceObjectRdf(
                         links, checksums, requestBodyStream, extContent);
             }
         } else {
-            final Model model = httpRdfService.bodyToInternalModel(externalPath(), requestBodyStream,
-                requestContentType);
+            final Model model = httpRdfService.bodyToInternalModel(externalUri, requestBodyStream,
+                requestContentType, identifierConverter);
```
Comment (awoods) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: For consistency and safety, probably best to use the `identifierConverter()` method.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java (hunk 1 above)
Comment (awoods) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: Actually, I have an update that puts the OCFL storage in the tmp-dir... as intended. 
https://jira.lyrasis.org/browse/FCREPO-3162

The "more permanent location" can currently be defined by the system properties: https://github.com/fcrepo4/fcrepo4/pull/1599/files#diff-f2c25ea789161a21300d5fa03b8a3965R43-R45

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java [hunk 10]
```diff
@@ -330,32 +322,32 @@ private RdfStream getResourceTriples(final int limit, final FedoraResource resou
             // Additional server-managed triples about this resource
             // Mementos already have the server managed properties in the PROPERTIES category
             // since mementos are immutable and these triples are no longer managed
-             if (ldpPreferences.prefersServerManaged() && !resource.isMemento()) {
-                streams.add(getTriples(resource, SERVER_MANAGED));
+            if (ldpPreferences.prefersServerManaged() && !resource.isMemento()) {
+                //streams.add(getTriples(resource, SERVER_MANAGED));
```
[the remark this answers is not in the document]
Comment (whikloj) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: I think include/omit choices were supposed to occur in a service...can't remember which service. So I agree this needs a `TODO` but I think it could be removed.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 8 above)
Comment (whikloj) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: I think this might be more easily tackled once we can get a response from the server. But I'm not opposed if @dbernstein feels he can fix them all, it will make this a more complex PR.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java [hunk 11]
```diff
@@ -409,7 +411,8 @@ public Response createOrReplaceObjectRdf(
             //}
         }
 
-        final String fedoraId = identifierConverter().toInternalId(externalPath());
+        final String externalUri = this.uriInfo.getRequestUri().toString();
+        final String fedoraId = identifierConverter().toInternalId(externalUri);
```
Comment (whikloj) [inline fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: Just noting that we get the `externalPath` from the [`@PathParam`](https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java#L138) so adding the host and context back on is why I added the [`toDomain()`](https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/api/rdf/HttpIdentifierConverter.java#L169) method to the HttpIdentifierConverter. So these two lines could be
```
final String fedoraId = identifierConverter().toInternalId(identifierConverter().toDomain(externalPath()));
```
Otherwise we could remove the `externalPath()` setup and make either `externalUri` and/or `fedoraId` into a class variable as this will probably be needed.

File: fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/api/rdf/HttpResourceConverter.java [hunk 12]
```diff
@@ -96,8 +97,7 @@ protected FedoraResource doForward(final Resource resource) {
 
     @Override
     protected Resource doBackward(final FedoraResource resource) {
-            //@TODO Implement this
-            return null;
+        return ResourceFactory.createResource(resource.getId());
```
Comment (whikloj) [inline fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/api/rdf/HttpResourceConverter.java]: This class has been replaced by the `HttpIdentifierConverter` perhaps adding a `@deprecated` to this is now necessary?

File: fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java [hunk 13]
```diff
@@ -111,6 +113,9 @@ public static void touchModificationHeaders(final ResourceHeadersImpl headers, f
         }
         headers.setLastModifiedDate(instant);
         headers.setLastModifiedBy(userPrincipal);
+
+        final String stateToken = DigestUtils.md5Hex(instant.toEpochMilli() + "").toUpperCase();
```
Comment (whikloj) [inline fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java]: I'm not 100% sure, but if this is casting the long to a String I think 
```
final String stateToken = DigestUtils.md5Hex(String.valueOf(instant.toEpochMilli())).toUpperCase();
```
would be clearer...unless it doesn't work as expected?

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java (hunk 1 above)
Comment (whikloj) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: I have merged FCREPO-3162, so this will need a rebase to pick up the changes.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: will do.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 8 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: Defer to JIRA:  https://jira.lyrasis.org/browse/FCREPO-3163

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java (hunk 1 above)
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: okay - I will not make any changes here.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java [hunk 14]
```diff
@@ -74,10 +74,10 @@
 import static org.fcrepo.kernel.api.RdfLexicon.VERSIONING_TIMEMAP_TYPE;
 import static org.fcrepo.kernel.api.RdfLexicon.isManagedNamespace;
 import static org.fcrepo.kernel.api.RdfLexicon.isManagedPredicate;
-import static org.fcrepo.kernel.api.RequiredRdfContext.EMBED_RESOURCES;
-import static org.fcrepo.kernel.api.RequiredRdfContext.INBOUND_REFERENCES;
-import static org.fcrepo.kernel.api.RequiredRdfContext.LDP_CONTAINMENT;
-import static org.fcrepo.kernel.api.RequiredRdfContext.LDP_MEMBERSHIP;
+//import static org.fcrepo.kernel.api.RequiredRdfContext.EMBED_RESOURCES;
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: will do.

File: fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java (hunk 13 above)
Comment (dbernstein) [reply fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java]: I'm not sure if casting a long to a string will cause the result to include commas.

File: fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/api/rdf/HttpResourceConverter.java (hunk 12 above)
Comment (dbernstein) [reply fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/api/rdf/HttpResourceConverter.java]: will add @Deprecated

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java (hunk 11 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: To be addressed post PR.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 8 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: I created a JIRA: https://jira.lyrasis.org/browse/FCREPO-3163

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 7 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: will do.

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java [hunk 15]
```diff
@@ -178,7 +181,10 @@ public boolean hasType(final String type) {
     @Override
     public RdfStream getTriples() {
         try {
-            return getSession().getTriples(id, getMementoDatetime());
+            final var triples = Stream.concat(getSession().getTriples(id, getMementoDatetime()),
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java]: yes - will do.

File: fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java (hunk 13 above)
Comment (awoods) [reply fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java]: I believe the "state token" should relate to the state of the resource, and therefore be reproducible in the case that the state of the resource does not change. 
The idea is that the state token can be subsequently used by a client when requesting to make a change to a resource. If the state token can not be reproduced on the server-side when determining whether the client-provided state token represents the current state or not... then the utility of the token is lost.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java (hunk 2 above)
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java]: This does not cover the case of an empty body: https://jira.lyrasis.org/browse/FCREPO-3168

File: fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java (hunk 13 above)
Comment (dbernstein) [reply fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java]: I think this should be addressed in a separate JIRA: https://jira.lyrasis.org/browse/FCREPO-3169

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java [hunk 16]
```diff
@@ -175,14 +176,25 @@ public boolean hasType(final String type) {
 
     @Override
     public List<URI> getTypes() {
+        if(types == null){
```
Comment (awoods) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java]: Minor: space after `if`

File: fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java [hunk 17]
```diff
@@ -49,18 +49,18 @@
  * @author whikloj
  * @since 2015-10-30
  */
-@Component
+@Provider
```
Comment (awoods) [inline fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: `@Component` is for Spring creation/scanning. `@Provider` is for jax-rs resources. Was this change intentional? ...it appears to have broken the auto-injection functionality.

Comment (dbernstein) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: I don't think it was intentionally - will fix.

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java (hunk 16 above)
Comment (dbernstein) [reply fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java]: will fix.

File: fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java (hunk 17 above)
Comment (dbernstein) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: Yes - it was intentional.  This provider uses the jax-rs @Context parameter for injecting resources.   I changed it to @Provider to fix the problem which was that the dependencies weren't being injected.

Comment (awoods) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: I just double-checked to make sure...
As far as I can tell, this change removes `LinkHeaderProvider` from any execution flow.
The previous logic was that `ContentExposingResource` would invoke the `HttpHeaderInjector`:
https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java#L691

..which looks for Spring-managed classes that implement `UriAwareHttpHeaderFactory`:
https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/api/HttpHeaderInjector.java#L81

By changing `LinkHeaderProvider`s annotation from `@Component` (Spring) to `@Provider` (jax-rs), the class is no longer found by `HttpHeaderInjector`... and therefore never invoked.

Comment (awoods) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: As a side note, this class only uses the `@Context` annotation for injecting resources because you made that change in this PR :)

Comment (awoods) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: There is a more fundamental issue, which may have been the reason for your introduction of this change, which how the `TransactionProvider` is now trying to be used. It is trying to be managed by both jax-rs and Spring:
https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/session/TransactionProvider.java#L38-L40

..which I do not think is possible. 

Additionally, the `TransactionProvider` implements `org.glassfish.hk2.api.Factory<Transaction>` in order to facilitate @Injecting Transactions into classes via HK2 (https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/ioc.html#d0e17033). It appears that an attempt was made to change this approach in https://github.com/fcrepo4/fcrepo4/pull/1559/files . Instead of Injecting `TransactionProvider`, I believe we should be Injecting `Transaction`... and allowing the TransactionProvider Factory to "provide" the Transaction.

I will look into...

Comment (awoods) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: Issues should now be resolved with: https://github.com/dbernstein/fcrepo4/pull/8

File: fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java (hunk 13 above)
Comment (bbpennel) [reply fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java]: Either `String.valueOf` or `Long.toString` would be fine and not add commas, but they might not matter at this point

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java [hunk 18]
```diff
@@ -98,7 +99,9 @@ public DefaultOCFLObjectSessionFactory(final File ocflStagingDir, final File ocf
 
     @Override
     public OCFLObjectSession create(final String ocflId, final String persistentStorageSessionId) {
-        final File stagingDirectory = new File(this.ocflStagingDir, persistentStorageSessionId);
+
+        final File stagingDirectory = new File(this.ocflStagingDir,
+                persistentStorageSessionId == null ? "read-only" : persistentStorageSessionId);
```
Comment (bbpennel) [inline fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: is "read-only" going to be referenced again in the future or is it just for the sake of having a string here? Wondering if it should be a constant.

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java (hunk 2 above)
Comment (bbpennel) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java]: Okay, what scenario do you get an null `triples` stream in a call to writeRDF?

Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/OCFLPersistentStorageUtils.java]: It looks like CreateRdfSourceOperationBuilder passes a null tripleStream when the stream from the request is empty.  So the resulting CreateRdfSourceOperation passes the null triples along.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java [hunk 19]
```diff
@@ -315,47 +303,54 @@ private RdfStream getResourceTriples(final int limit, final FedoraResource resou
 
         final List<Stream<Triple>> streams = new ArrayList<>();
 
-
         if (returnPreference.getValue().equals("minimal")) {
-            streams.add(getTriples(resource, of(PROPERTIES, MINIMAL)).filter(tripleFilter));
-
+            streams.add(resource.getTriples().filter(tripleFilter));
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: I'll remove.

File: fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/ManagedPropertiesService.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+ */
+package org.fcrepo.kernel.api.services;
+
+import org.apache.jena.graph.Triple;
+import org.fcrepo.kernel.api.models.FedoraResource;
+
+import java.util.stream.Stream;
+
+/**
+ * Interface for a service to create a new resource via a POST request.
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/ManagedPropertiesService.java]: I will fix.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java (hunk 11 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: I'll updated it according to your suggestion.

Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: @whikloj :  I'd prefer to defer this change for another PR in the interest of moving things along.  Would you mind creating a JIRA describing the changes you are recommending?

File: fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java (hunk 13 above)
Comment (dbernstein) [reply fcrepo-persistence-common/src/main/java/org/fcrepo/persistence/common/ResourceHeaderUtils.java]: Done.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java (hunk 11 above)
Comment (whikloj) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: https://jira.lyrasis.org/browse/FCREPO-3170

File: fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java (hunk 18 above)
Comment (dbernstein) [reply fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionFactory.java]: Ah - yeah:  I put that there for the sake of having a string - it won't be referenced elsewhere.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 8 above)
Comment (whikloj) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: I'm good with this, @awoods can we consider this *resolved*?

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java (hunk 9 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java]: This is done.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 7 above)
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: Done.

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: Done

File: fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (dbernstein) [reply fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java]: Done

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java (hunk 16 above)
Comment (dbernstein) [reply fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/FedoraResourceImpl.java]: Done

File: fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java (hunk 17 above)
Comment (dbernstein) [reply fcrepo-auth-webac/src/main/java/org/fcrepo/auth/webac/LinkHeaderProvider.java]: Merged.
